Variadic ODE

rok-cesnovar opened the pull request:
Fixes #481

rok-cesnovar commented:
@bbbales2 once https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3-test-binaries/detail/stanc3-test-binaries/87/pipeline finishes

put this in make/local

STANC3_TEST_BIN_URL = https://jenkins.mc-stan.org/job/stanc3-test-binaries/87/artifact

rm bin/stanc and make build.

The old interface doesnt work right now but the variadic does.

Also, it treats 5e2 as real so this is valid

 vector[2] z[N]
  = ode_bdf_tol(dz_dt, z_init, 0, ts,
		1e-5, 1e-3, 500,
		alpha, beta, gamma, delta);

this isnt

 vector[2] z[N]
  = ode_bdf_tol(dz_dt, z_init, 0, ts,
		1e-5, 1e-3, 5e2,
		alpha, beta, gamma, delta);

rok-cesnovar commented:
Lord Jenkins was not pleased the first time. The stanc3 binaries are now ready with the magic number 88. 

So use:
STANC3_TEST_BIN_URL = https://jenkins.mc-stan.org/job/stanc3-test-binaries/88/artifact

rok-cesnovar commented:
@bbbales2 @wds15 the _tol signatures should probably not allow the scientific notation of max steps. This will not be valid:
```
vector[2] z[N]
  = ode_bdf_tol(dz_dt, z_init, 0, ts,
		1e-5, 1e-3, 5e3,
		alpha, beta, gamma, delta);
```
The alternative is to handle this at the C++ level somehow also. See https://github.com/stan-dev/stanc3/issues/596

wds15 commented:
fine for me to not allow this (I guess to make it integer).

bbbales2 commented:
Yeah it's totally fine.

wds15 commented:
@rok-cesnovar does this PR display deprecation messages when the parser finds any of the old integrate* functions in a Stan file? I think we should do that as the older function will be slightly slower than the new ones due to the need to use adapters.

rok-cesnovar commented:
They do not, because we do not currently have the deprecation warnings for functions working (outside of the parser level ones (# for comment, <- for assignments, etc.).

There are 3 issues for that open (https://github.com/stan-dev/stanc3/issues/600, https://github.com/stan-dev/stanc3/issues/165 and https://github.com/stan-dev/stanc3/issues/261) and this seems like something simple to implement (to my naive mind) so maybe we can add this in the meantime while waiting you guys to hash out the math PR.

rok-cesnovar commented:
@bbbales2 

Just double checking. In your test models you have

zd = ode_rk45_tol(f_1_arg, vd, rd, rad, 1e-6, 1e-6, 100, rd);
zd = ode_rk45_tol(f_0_arg, vd, rd, rad, 1e-6, 1e-6, 100);

but those should not be allowed right? I was under the impression that two variables follow the max_num_steps?
Or am I missing something?

bbbales2 commented:
> but those should not be allowed right? I was under the impression that two variables follow the max_num_steps?
Or am I missing something?

Those are all the pass through arguments. It makes sense to pass through zero or one argument.

rok-cesnovar commented:
Oh ok. Thanks. Will fix.

rok-cesnovar commented:
This is ready for review. Tests will only pass once Math PR is in but the review does not have to wait.

bbbales2 commented:
@rok-cesnovar just asked a question over [here](https://github.com/stan-dev/stan/pull/2937#pullrequestreview-446578416) that's relevant

rybern commented on src/frontend/Semantic_check.ml:
```diff
@@ -347,6 +348,92 @@ let semantic_check_reduce_sum ~is_cond_dist ~loc id es =
       |> Validate.error
 
+let semantic_check_variadic_ode_tol ~is_cond_dist ~loc id es =
+  let arg_match (x_ad, x_t) y =
```
There's some code replication between this function, semantic_check_variadic_ode, and semantic_check_reduce_sum. I think we can at minimum pull out arg_match and args_match

rok-cesnovar replied:
Pulled arg match and args_match out of the semantic checks and collapsed the separate _tol and normal check.

rybern commented on src/frontend/Semantic_check.ml:
```diff
@@ -347,6 +348,92 @@ let semantic_check_reduce_sum ~is_cond_dist ~loc id es =
+             :: rel_tol
+                :: abs_tol
+                   :: { emeta= {type_= UInt; ad_level= UnsizedType.DataOnly; _}; _
+                      } (* max_num_steps *)
```
Ideally, we would encapsulate the information about the signature of the ODEs inside stan_math_signatures. I know it's not obvious how to do that until someone takes the time to make the signatures flexible enough to encode these variadic functions, but maybe some of this signature-specific information can be pulled out?

rok-cesnovar replied:
Will try to add something to signatures. Seems reasonable yeah.

rybern commented on src/frontend/Semantic_check.ml:
```diff
@@ -347,6 +348,92 @@ let semantic_check_reduce_sum ~is_cond_dist ~loc id es =
+         && arg_match (AutoDiffable, UVector) initial_state
+         && arg_match (AutoDiffable, UArray UReal) times
+         && arg_match (DataOnly, UReal) rel_tol
+         && arg_match (DataOnly, UReal) abs_tol ->
```
It seems like these two lines and error types are the only difference between this function and semantic_check_variadic_ode? Can we call that one here, or have them share a helper function? That'd save ~50 lines of replicated code

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +162,96 @@ module TypeError = struct
+          generate_ode_sig
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          arg_tys
+    | IllTypedVariadicODEGenericTol (name, arg_tys) ->
```
Again I'm wondering if there's a less redundant way to write this, since it only differs from IllTypedVariadicODEGeneric by the list of types - but it might not be worth it here, up to you

rok-cesnovar replied:
Will try to reduce the redundancy. Good call

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +162,96 @@ module TypeError = struct
+        in
+        Fmt.pf ppf
+          "Ill-typed arguments supplied to function '%s'. Available arguments:\n\
+           %sWhere T1 and T2 are of any type.@[<h>Instead supplied arguments \
```
do we still have this minimum of two types, T1 and T2?

rok-cesnovar replied:
No. Missed this one. Thanks.

rybern commented on src/middle/Stan_math_signatures.ml:
```diff
@@ -140,8 +140,26 @@ let mk_declarative_sig (fnkinds, name, args) =
 let full_lpmf = [Lpmf; Rng; Ccdf; Cdf]
 let reduce_sum_functions = ["reduce_sum"; "reduce_sum_static"]
+
+let variadic_ode_functions =
```
tiny nitpick - these might be slightly faster as a set, since the only operation seems to be checking membership

rok-cesnovar replied:
Thanks. Will move to set!

rybern commented on src/stan_math_backend/Stan_math_code_gen.ml:
```diff
@@ -775,6 +797,21 @@ let pp_register_map_rect_functors ppf p =
     (list ~sep:cut pp_register_functor)
     (List.sort ~compare (Hashtbl.to_alist map_rect_calls))
 
+let fun_used_in_ode_bdf p =
```
this function seems the same as `fun_used_in_reduce_sum` except for the function matching predicate. I'd suggest making a local helper function and pulling out that predicate as an argument

rybern commented:
Overall looks quite good to me! Mostly just some small suggestions.

It would be best not to have to handle this as a special case in semantic check, just like reduce_sum, but someone would need to make the stan math type signatures flexible enough to represent these variatic functions.

wds15 commented:
@rok-cesnovar How can I grab the latest stanc3 pre-build binary? What URL do I need? I wanted to rerun the SIR examples, just to make sure that using the new interface gives the right performance. Thanks.

rok-cesnovar commented:
The latest binary is still https://jenkins.mc-stan.org/job/stanc3-test-binaries/97/

But that one only works for the new interface and not for the old one (we reverted to the old functor). I am doing that update together with addressing the review comments. I am in the middle of fixing that but probably wont finish until the afternoon. 

So for now you can use the new interface with this binary and the old binary with cmdstan develop. Not ideal, I know.

wds15 commented:
Ah... ok.. so I can at least use the 97 to process a variadic ODE Stan model... will do that.

rok-cesnovar commented:
Yes, for variadic ODE 97 is fine. The only things that changed since 97 for the variadic interface are cosmetic changes to the OCaml code, the generated code is still the same.

rok-cesnovar commented:
And the code-gen will stay the same, so these tests that you will perform will be valid for the final version too. All the requested changes are cosmetic or touch the old interface.

rok-cesnovar commented:
I addressed all review comments, moved all signature types related info to stan math signatures and reduced the code duplication. This passes tests now https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3/detail/PR-525/43/pipeline but Github is having some issues ATM so its not showing up.

wds15 commented:
@rok-cesnovar This could be a bit late... but we were just talked over it at the Stan Math meeting (with @bbbales2 @SteveBronder and @syclik ) ... the issue with variadic ODE is that it slows down a lot the old integrate_ode* stuff. So what about if the stanc3 compiler translates the calls of the old intrgrate_ode* directly to the new variadic ODE interface? This could be done by emitting respective C++ code or by mapping things within the Stan language (which is hopefully the easier way with all the AST magic you have).

If not doable for now... then ideally in a little bit. This would allow all existing code to keep running at full speed without any user change to their models.

rok-cesnovar commented:
Sorry I missed the meeting. I think this is doable and a great idea. Let me give it a try.

rok-cesnovar commented:
So we will have to change the input y0 to a vector and the result from an array of vectors to a 2d array. @wds15 @bbbales2 I am guessing this will hurt performance but no to a 20% extent?

wds15 commented:
It won’t affect performance at all if you just do that conversion once. The issue with the current adaptier is that it’s done every time when the ode rhs is called, but you should be able to code it up such that it’s done a single time such that there is no penalty.

rok-cesnovar commented:
Oh, so currently the conversion is done for each step in the solve?

wds15 commented:
Yes

bbbales2 commented:
> should be able to code it up such that it’s done a single time such that there is no penalty.

@wds15 I think we always have to cast the output. We'd have to modify the internals of the function to make it return a vector.

wds15 commented:
But the parser can just create the old and a converted ode rhs code in order to avoid the conversion upon each call.

I hope it’s that easy...it does sound easy to me, but I don’t deal with the parser...

wds15 commented:
Ah...I see...the issue is that the user function must be modified...hmm.

bbbales2 commented:
I don't think it will be easy to convert all the variations in how stuff can be written:

```
real out[1];

out[1] = theta[1] * y[1];

return out;
```

```
real out[1] = { theta[1] * y[1] };
return out;
```

```
return { theta[1] * y[1] };
```

to make them vector-like. I think the difficulty of automatically converting the output to be a vector would be very high. The input should be do-able though, and then if we just convert the output that's 50% of the losses gone.

bbbales2 commented:
Yeah yeah, but I think the input state could still be handled. We'll see. This would have been better if we did it earlier, but deprecation warnings are good enough (I'm very happy to just stay with what we've got if that means a smooth release).

wds15 commented:
Unless the parser can do magic with the ast, mir or anything like that we should probably let this alone. The idea is sweat, but dies on the details.

rok-cesnovar commented:
No magic yeah. So just a review left to do. Attaching the Linux binary. unfortunately, I cant build on Jenkins right now, some Windows machine issues.

[stanc.zip](https://github.com/stan-dev/stanc3/files/4933847/stanc.zip)

wds15 commented:
Is this still on track for tomorrow? Would be good to get it in 2.24... as we slow down ODEs by 30% otherwise...

rok-cesnovar commented:
The first review comments were addressed so just a final review needed.

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
           (String.concat ~sep:"" lines)
           Fmt.(list UnsizedType.pp ~sep:comma)
           arg_tys
+    | IllTypedVariadicODE (name, arg_tys, args) ->
```
I think this could still have been two errors like you had it, I only meant to suggest that they could both use a helper function because their logic overlapped so much. But this works too, I'm sorry if I put you through extra effort

rok-cesnovar replied:
No problem, I think this is more concise and also makes less code to maintain or fix when we add the type for variadic functions.

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
           Fmt.(list UnsizedType.pp ~sep:comma)
           arg_tys
+    | IllTypedVariadicODE (name, arg_tys, args) ->
+        let types x = List.map ~f:(fun (_, t) -> t) x in
```
small note: `(fun (_, t) -> t)` is also called `snd`

rok-cesnovar replied:
Thanks. Fixed.

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
+        let optional_tol_args =
+          if Stan_math_signatures.is_variadic_ode_tol_fn name then
+            List.map
+              ~f:(fun (_, y) -> y)
```
could use your `types` function here?

rok-cesnovar replied:
Yeah, good call. Done.

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
+              Stan_math_signatures.variadic_ode_tol_arg_types
+          else []
+        in
+        let args_types = List.map ~f:(fun (_, t) -> t) args in
```
again could use `types` function

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
+          @ types Stan_math_signatures.variadic_ode_mandatory_arg_types
+          @ optional_tol_args @ args_types
+        in
+        let variadic_ode_generic_signature =
```
IMO it's not clear why you're building these things, a comment describing your strategy throughout this section might make it more clear

rok-cesnovar replied:
Added this comment: 

This function is used to generate the generic signature for variadic ODEs, i.e. with ... representing the variadic parts of the signature. This should be removed once a type representing variadic arguments is added. The generic signature is printed when there is a semantic error with one of the non-variadic arguments in the signature. If there is a mismatch in the variadic arguments, we print the non-generic expected signature (with explicit types for variadic args).

rybern commented on src/frontend/Semantic_error.ml:
```diff
@@ -152,6 +156,68 @@ module TypeError = struct
+          in
+          let nth_arg n =
+            List.nth_exn
+              (types Stan_math_signatures.variadic_ode_mandatory_arg_types)
```
this might recompute the result of `(types ...)` every time it's called - very tiny issue, but FWIW one alternative way of writing it might be something like

```
let arg0 :: arg1 :: arg2 :: _ = types ...
```

rok-cesnovar replied:
Thanks. But this way I get a warning for a not-exhaustive pattern matching that I am unable to fix: 
File "src/frontend/Semantic_error.ml", line 192, characters 10-471:
Error (warning 8): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(_::_::[]|_::[]|[])

rok-cesnovar replied:
Ignore the above. Fixed.

rybern replied:
Oh, yeah, that's unfortunate. You could get around it like:

```
let (arg0, arg1, arg2) = match types ... with
   a0 :: a1 :: a2 :: _ -> (a0, a1, a2)
   _ -> error "oops"
```

but again it's a very small issue that probably doesn't deserve our attention :P

rok-cesnovar replied:
Went with:

```ocaml
match
            ( types Stan_math_signatures.variadic_ode_mandatory_arg_types
            , types Stan_math_signatures.variadic_ode_mandatory_fun_args )
          with
          | arg0 :: arg1 :: arg2 :: _, fun_arg0 :: fun_arg1 :: _ ->
              Fmt.strf "(%a, %a, ...) => %a, %a, %a, %a, %a ...\n"
                UnsizedType.pp fun_arg0 UnsizedType.pp fun_arg1 UnsizedType.pp
                Stan_math_signatures.variadic_ode_fun_return_type
                UnsizedType.pp arg0 UnsizedType.pp arg1 UnsizedType.pp arg2
                Fmt.(list UnsizedType.pp ~sep:comma)
                optional_tol_args
          | _ ->
              raise_s
                [%message
                  "This should not happen. Variadic ODE functions have exactly three \
                   mandatory arguments and the function supplied to the variadic ODE \
                   function has exactly two mandatory arguments."]
```

rybern replied:
Looks good!

rybern commented on src/stan_math_backend/Stan_math_code_gen.ml:
```diff
@@ -242,7 +255,16 @@ let pp_fun_def ppf Program.({fdrt; fdname; fdargs; fdbody; _})
-                "Ill-formed reduce_sum call! This is bug in the compiler."] )
+                "Ill-formed reduce_sum call! This is bug in the compiler."]
+      else if String.Set.mem funs_used_in_ode_bdf fdname then
+        (* Produces the reduce_sum functors that has the pstream argument
```
outdated comment here i think

rok-cesnovar replied:
Indeed. Fixed.

rok-cesnovar commented:
Thanks @rybern ! 

I addressed all your comments. If you are good with the pattern matching thing on args, I will merge once tests pass.

rybern commented:
Looks good to me
